### ISSUE TITLE
Add `haskey` for `Tuple`

### ISSUE THREAD

**Issue by simonschoelly** [JuliaLang/julia#35516]
Currently `haskey` is defined for `NamedTuple` but not for `Tuple`. Would it make sense to also add this for tuples? After all, `keys(::Tuple)` is also defined.

Code would then probably look like this
```julia
    haskey(t::Tuple, key) = isdefined(t, key)
```
Related: https://github.com/JuliaLang/julia/issues/32898

**Comment by timholy** [comment]
Or just a generic fallback `haskey(container, key) = key ∈ keys(container)`, if we decide that `keys` is the core trait/interface function.

For tuples, at least, it's efficient:
```julia
julia> hk(container, key) = key ∈ keys(container)
hk (generic function with 1 method)

julia> t = (5, 7)
(5, 7)

julia> @code_llvm hk(t, 2)

;  @ REPL[21]:1 within `hk'
define i8 @julia_hk_17332([2 x i64] addrspace(11)* nocapture nonnull readonly dereferenceable(16), i64) {
top:
; ┌ @ range.jl:997 within `in'
; │┌ @ bool.jl:40 within `&'
    %.off = add i64 %1, -1
    %2 = icmp ult i64 %.off, 2
; └└
  %3 = zext i1 %2 to i8
  ret i8 %3
}
```

**Comment by simeonschaub** [comment]
+1 from me. We should probably also define `hasproperty`. I think this would only be consistent, since we currently have:
```
julia> keys((1,))
Base.OneTo(1)

julia> propertynames((1,))
(1,)

julia> haskey((1,), 1)
ERROR: MethodError: no method matching haskey(::Tuple{Int64}, ::Int64)
Closest candidates are:
  haskey(::Base.Iterators.Pairs, ::Any) at iterators.jl:267
  haskey(::WeakKeyDict{K, V} where V, ::Any) where K at weakkeydict.jl:111
  haskey(::Base.ImmutableDict, ::Any) at dict.jl:772
  ...
Stacktrace:
 [1] top-level scope
   @ REPL[3]:1

julia> hasproperty((1,), 1)
ERROR: MethodError: no method matching hasproperty(::Tuple{Int64}, ::Int64)
Closest candidates are:
  hasproperty(::Any, ::Symbol) at reflection.jl:1545
Stacktrace:
 [1] top-level scope
   @ REPL[4]:1
```
A generic fallback might be a good idea though.

**Comment by puerco-errante** [comment]
Hi! I'm looking for some simple issues to start collaborating in Julia. Could you point me to some general guidelines for how to submit a pull request? @timholy suggestion seems quite general, although I'm a bit confused about the aim of this. For tuples one could use something along the lines of 

 ```julia
haskey(t::Tuple, key) = False
haskey(t::Tuple, key::Int) = key<length(t)
```

In any case, I'd be happy to code whichever option seems more Julia-friendly, just let me know which are the steps.

Cheers and thanks in advance!

**Comment by simeonschaub** [comment]
Check out https://kshyatt.github.io/post/firstjuliapr/, which should give you a pretty good idea for how to get started. I think just defining `haskey(t::Tuple, key::Int)` would be fine for now, we can always worry about a more general solution later. Your implementation isn't quite correct for indices smaller than 1 and for `key == length(t)`. `key in keys(t)` should compile down to the same code as manually checking if `key` is between `1` and `length(t)`, so I would probably just use that.

**Comment by puerco-errante** [comment]
You're right. I keep trying to outsmart the compiler to no avail, heheh. 
I'll give a look the link you mentioned and hopefully implement something that works soon.

**Comment by simonschoelly** [comment]
One can also index into a tuple using any `Real`, so I wonder if haskey should then also work for all subtypes of `Real` and not just `Int`. This is a bit inconsistend: `NamedTuple` can only be indexed by `Int`.

**Comment by puerco-errante** [comment]
Indeed,
> One can also index into a tuple using any `Real`, so I wonder if haskey should then also work for all subtypes of `Real` and not just `Int`. This is a bit inconsistend: `NamedTuple` can only be indexed by `Int`.
using
```julia
haskey(t::Tuple, key) = isdefined(t, key)
haskey((1,2,3),0.0)
```
gives an error.

`hk(container, key) = key ∈ keys(container)` on the other hand seems to work alright.
